lib/mongo/auth/aws/credentials_retriever.rb

```diff
@@ -0,0 +1,181 @@
+# Copyright (C) 2020 MongoDB Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Mongo
+  module Auth
+    class Aws
+
+      # Retrieves AWS credentials from a variety of sources.
+      #
+      # This class provides for AWS credentials retrieval from:
+      # - the passed user (which receives the credentials passed to the
+      #   client via URI options and Ruby options)
+      # - AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_SESSION_TOKEN
+      #   environment variables (commonly used by AWS SDKs and various tools,
+      #   as well as AWS Lambda)
+      # - EC2 metadata endpoint
+      # - ECS metadata endpoint
+      #
+      # The sources listed above are consulted in the order specified.
+      # The first source that contains any of the three credential components
+      # (access key id, secret access key or session token) is used.
+      # The credential components must form a valid set if any of the components
+      # is specified; meaning, access key id and secret access key must
+      # always be provided together, and if a session token is provided
+      # the key id and secret key must also be provided. If a source provides
+      # partial credentials, credential retrieval fails with an exception.
+      #
+      # @api private
+      class CredentialsRetriever
+
+        # Timeout for metadata operations, in seconds.
+        #
+        # The auth spec suggests a 10 second timeout but this seems
+        # excessively long given that the endpoint is essentially local.
+        METADATA_TIMEOUT = 5
```

Comment:
We should stick to the spec and use the 10 second timeout for consistency across drivers. Feel free to open a drivers ticket to reduce the timeout to 5 seconds.

Reply:
The spec suggests a 10 second timeout but does not require it.

Reply:
Created https://jira.mongodb.org/browse/DRIVERS-994.

Reply:
In our specs "SHOULD" usually means "MUST unless it's impossible or overly complex to implement in the lanaguage/driver". But reading the rationale in the ticket I agree we should reduce it to 5, thanks for opening.